File: server-go/internal/input/input.go
package input

import "strings"

type State struct {
	Input   string
	Hist    int // -1 = none
	Cmd     string
	Picking bool
	Pick    int
}

type Action struct {
	Kind   string
	Key    string
	Text   string
	Index  int
	Repeat bool // safe to fire again while the key is held
}

type Result struct {
	State  State
	Action Action
}

type KeyCtx struct {
	Awaiting bool
	History  []string
	Beacons  int
}

func isDigits(text string) bool {
	if text == "" {
		return false
	}
	for _, c := range text {
		if c < '0' || c > '9' {
			return false
		}
	}
	return true
}

const CommandPrefix = ";"

var arrows = map[string]bool{"up": true, "down": true, "left": true, "right": true}

// suggest returns the newest history entry that extends the current input, or
// "" when there is none — the basis for autosuggest + Tab-accept.
func Suggest(input string, history []string) string {
	if input == "" {
		return ""
	}
	for i := len(history) - 1; i >= 0; i-- {
		if history[i] != input && strings.HasPrefix(history[i], input) {
			return history[i]
		}
	}
	return ""
}

func splitMod(k string) (mod, base string) {
	i := strings.IndexByte(k, '+')
	if i <= 0 {
		return "", k
	}
	return k[:i], k[i+1:]
}

func commandKey(state State, key string) Result {
	typing := func(cmd string) Result {
		return Result{State{Input: state.Input, Hist: state.Hist, Cmd: cmd}, Action{Kind: "none"}}
	}
	if key == "esc" {
		return typing("")
	}
	if key == "enter" {
		return Result{State{Input: state.Input, Hist: state.Hist}, Action{Kind: "command", Text: strings.TrimPrefix(state.Cmd, CommandPrefix)}}
	}
	if key == "backspace" {
		return typing(state.Cmd[:len(state.Cmd)-1])
	}
	if len(key) == 1 {
		return typing(state.Cmd + key)
	}
	return typing(state.Cmd)
}

const PickerKey = "fn+esc"

func pickerKey(state State, key string, ctx KeyCtx) Result {
	stay := func(pick int) Result {
		return Result{State{Input: state.Input, Hist: state.Hist, Picking: true, Pick: pick}, Action{Kind: "none"}}
	}
	leave := func(a Action) Result {
		return Result{State{Input: state.Input, Hist: state.Hist}, a}
	}
	if key == "esc" || key == PickerKey {
		return leave(Action{Kind: "none"})
	}
	if ctx.Beacons <= 0 {
		return stay(state.Pick)
	}
	if key == "up" || key == "down" {
		delta := 1
		if key == "up" {
			delta = -1
		}
		return stay(((state.Pick+delta)%ctx.Beacons + ctx.Beacons) % ctx.Beacons)
	}
	if key == "enter" {
		return leave(Action{Kind: "connect", Index: state.Pick})
	}
	if len(key) == 1 && key[0] >= '1' && key[0] <= '9' {
		return leave(Action{Kind: "connectRow", Index: int(key[0] - '1')})
	}
	return stay(state.Pick)
}

func InterpretKey(state State, key string, ctx KeyCtx) Result {
	if state.Picking {
		return pickerKey(state, key, ctx)
	}
	if key == PickerKey {
		return Result{State{Input: state.Input, Hist: state.Hist, Picking: true}, Action{Kind: "none"}}
	}
	if state.Cmd != "" {
		return commandKey(state, key)
	}
	input := state.Input
	hist := state.Hist
	mirror := func(newInput string, newHist int) State { return State{Input: newInput, Hist: newHist} }
	press := func(s State, k string) Result {
		return Result{s, Action{Kind: "pressKey", Key: k, Repeat: arrows[k]}}
	}
	quiet := func(s State) Result { return Result{s, Action{Kind: "none"}} }

	mod, base := splitMod(key)

	if mod == "shift" {
		if base == "esc" {
			return press(state, "escape")
		}
		if base == "enter" {
			return quiet(mirror(input+"\n", hist))
		}
		return quiet(state)
	}
	if mod == "ctrl" {
		if base == "=" || base == "+" {
			return Result{state, Action{Kind: "zoom", Key: "in"}}
		}
		if base == "-" || base == "_" {
			return Result{state, Action{Kind: "zoom", Key: "out"}}
		}
		if base == " " {
			return Result{state, Action{Kind: "zoom", Key: "reset"}}
		}
		if base == "up" {
			if len(ctx.History) == 0 {
				return quiet(state)
			}
			idx := len(ctx.History) - 1
			if hist != -1 {
				idx = max(hist-1, 0)
			}
			return quiet(mirror(ctx.History[idx], idx))
		}
		if base == "down" {
			if hist == -1 {
				return quiet(state)
			}
			idx := hist + 1
			if idx >= len(ctx.History) {
				return quiet(mirror("", -1))
			}
			return quiet(mirror(ctx.History[idx], idx))
		}
		if len(base) == 1 {
			return press(state, "ctrl+"+base)
		}
		return quiet(state)
	}
	if mod == "opt" {
		if arrows[base] {
			return press(state, base)
		}
		return quiet(state)
	}

	if key == "esc" {
		if len(input) > 0 {
			return quiet(mirror("", -1))
		}
		return press(state, "escape")
	}
	if key == "enter" {
		if len(input) > 0 {
			return Result{mirror("", -1), Action{Kind: "send", Text: input}}
		}
		return press(state, "enter")
	}
	if key == "backspace" {
		if input == "" {
			return quiet(mirror("", hist))
		}
		return quiet(mirror(input[:len(input)-1], hist))
	}
	if key == "tab" {
		if s := Suggest(input, ctx.History); s != "" {
			return quiet(mirror(s, -1))
		}
		return press(state, "tab")
	}
	if arrows[key] {
		return Result{state, Action{Kind: "pan", Key: key, Repeat: true}}
	}
	if len(key) == 1 {
		if input == "" && key == CommandPrefix {
			return quiet(State{Input: input, Hist: hist, Cmd: CommandPrefix})
		}
		if ctx.Awaiting && input == "" && isDigits(key) {
			return press(state, key)
		}
		return quiet(mirror(input+key, hist))
	}
	return quiet(state)
}

File: server/lib/input.js
'use strict';

// Pure interaction FSM — the LAZY-DEVICE contract. The device forwards every raw
// key; the SERVER owns the input buffer, all special functions and what reaches
// the terminal. Keys and actions use GENERIC names only — no backend (tmux)
// spelling here; the terminal adapter translates. No regex, no else.
//
// State: { mode: 'mirror', input: string, hist: number|null }
// Keys:  single chars, arrows (up/down/left/right), enter, backspace, tab, esc,
//        and modifier chords as '<mod>+<base>' (shift+enter, shift+esc,
//        ctrl+<char>, ctrl+up/down, opt+<arrow>).
//
// Actions: none | send{text} | pressKey{key} | pan{key}
//
//   char        -> append to the input buffer (digit answers a menu when
//                  awaiting and the buffer is empty -> pressKey digit)
//   backspace   -> delete last char
//   shift+enter -> append '\n' (compose multi-line)
//   enter       -> send the buffer (empty buffer -> pressKey enter)
//   tab         -> pressKey tab (selector auto-advance / completion)
//   esc         -> typing? CLEAR the input : real Escape into the terminal
//   shift+esc   -> pressKey escape (STOP the agent; draft kept)
//   ctrl+up/dn  -> history recall prev / next (shell idiom)
//   ctrl+<char> -> pressKey ctrl+<char> (Ctrl+C & co.)
//   arrows      -> pan (read around; numbers choose menu options)
//   opt+<arrow> -> pressKey arrow (drive the terminal's own selector)

function isDigits(text) {
  if (!text) return false;
  for (const c of text) {
    if (c < '0' || c > '9') return false;
  }
  return true;
}

const ARROWS = new Set(['up', 'down', 'left', 'right']);

// Split '<mod>+<base>' once — the single modifier parse for the whole FSM.
function splitMod(k) {
  const i = k.indexOf('+');
  if (i <= 0) return { mod: '', base: k };
  return { mod: k.slice(0, i), base: k.slice(i + 1) };
}

const press = (state, key) => ({ state, action: { kind: 'pressKey', key } });
const quiet = (state) => ({ state, action: { kind: 'none' } });

function interpretKey(state, key, ctx) {
  const input = (state && state.input) || '';
  const hist = state && typeof state.hist === 'number' ? state.hist : null;
  const k = String(key == null ? '' : key);
  const awaiting = !!(ctx && ctx.awaiting);
  const history = (ctx && ctx.history) || [];
  const mirror = (newInput, newHist = null) => ({ mode: 'mirror', input: newInput, hist: newHist });

  // Parse any modifier chord exactly once.
  const { mod, base } = splitMod(k);

  if (mod === 'shift') {
    if (base === 'esc') return press(state, 'escape');                    // STOP the agent; draft kept
    if (base === 'enter') return quiet(mirror(input + '\n', hist));      // compose multi-line
    return quiet(state);
  }
  if (mod === 'ctrl') {
    if (base === 'up') {                                                  // history prev (shell idiom)
      if (history.length === 0) return quiet(state);
      const idx = hist === null ? history.length - 1 : Math.max(0, hist - 1);
      return quiet(mirror(history[idx], idx));
    }
    if (base === 'down') {                                                // history next
      if (hist === null) return quiet(state);
      const idx = hist + 1;
      if (idx >= history.length) return quiet(mirror(''));
      return quiet(mirror(history[idx], idx));
    }
    if (base.length === 1) return press(state, 'ctrl+' + base);           // Ctrl+C & co.
    return quiet(state);
  }
  if (mod === 'opt') {
    if (ARROWS.has(base)) return press(state, base);
    return quiet(state);
  }

  if (k === 'esc') {
    if (input.length > 0) return quiet(mirror(''));                       // clear (also drops recall)
    return press(state, 'escape');                                        // real Escape (ssh-parity)
  }
  if (k === 'enter') {
    if (input.length > 0) return { state: mirror(''), action: { kind: 'send', text: input } };
    return press(state, 'enter');
  }
  if (k === 'backspace') return quiet(mirror(input.slice(0, -1), hist));
  if (k === 'tab') return press(state, 'tab');
  if (ARROWS.has(k)) return { state, action: { kind: 'pan', key: k } };
  if (k.length === 1) {
    if (awaiting && input.length === 0 && isDigits(k)) return press(state, k);   // answer a menu
    return quiet(mirror(input + k, hist));
  }
  return quiet(state);
}

module.exports = { interpretKey, isDigits };
